### PR TITLE
Nested scatters and OGIN lookups

### PR DESCRIPTION
**Pull request by cjllanwarne**
- [x] Rebase onto develop after #3374 

### Sample WDL

```wdl
version draft-3

workflow nested_scatter {

  Array[Int] indices = [1,2,3]
  Int y = 55

  scatter(a in indices) {
    scatter(b in indices) {
      Int x = a + b
      scatter(c in indices) {
        Int j = a + b + c + x
      }
      scatter(d in j) {
        Int k = d + y
      }
    }
  }

  output {
    Array[Array[Array[Int]]] ks = k
  }
}
```

### Sample output
```json
{
  "nested_scatter.ks": [[[60, 61, 62], [62, 63, 64], [64, 65, 66]], [[62, 63, 64], [64, 65, 66], [66, 67, 68]], [[64, 65, 66], [66, 67, 68], [68, 69, 70]]]
}
```

### CODE REVIEW

**Comment by cjllanwarne** [comment]
I need a red thumb because of some changes in WOM - specifically I needed some way to indicate that a not-really-a subworkflow call should not get its subworkflow name prepended.

I'm also curious whether the "make up a random UUID for the workflow name" is going to cause problems down the road (eg does restart rely on stable subworkflow names?)

**Comment by Horneth** [comment]
> eg does restart rely on stable subworkflow names?

Yes - otherwise it would start sub-workflows again from scratch every time. There's a `SUB_WORKFLOW_STORE_ENTRY` table that maintains the relationship between a sub workflow, its parent, and the root workflow

**Comment by cjllanwarne** [comment]
Thanks @Horneth, I feared that might be the case!

I also have vague memories of scatters being numbered in draft 2 but can't remember why - do you remember what that was all about?

**Comment by Horneth** [comment]
Hmm I'm not totally sure - I think the full version of the Scope FQN includes something like `$scatter_n` where `n` is the "number" of the scatter. I think we use to persist the Scatter in the DB for some reason as a placeholder and we might have needed that for uniqueness if there are several scatters. Those are vague memories too so I'm not entirely sure.

**Comment by cjllanwarne** [comment]
@Horneth right, and we don't need to do that any more because we rebuild the entire execution store every time we restart.

I've updated this to now have stable subworkflow names for those subworkflows. Do we have an existing "check subworkflows restart correctly" test I can duplicate for draft 3?

**Comment by mcovarr** [comment]
Should nested scatters remain broken in pre-draft 3?

File: wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/wdlom2wom/graph/ScatterElementToGraphNode.scala [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+      val ogins: Set[GraphNode] = (foundOuterGenerators.toList map { case (name: String, port: OutputPort) =>
+        OuterGraphInputNode(WomIdentifier(name), port, preserveScatterIndex = false)
+      }).toSet
+
+      val graphLikeConvertInputs = GraphLikeConvertInputs(graphElements.toSet, ogins ++ Set(womInnerGraphScatterVariableInput), a.availableTypeAliases, a.workflowName, insideAScatter = true)
+      val innerGraph: ErrorOr[Graph] = WorkflowDefinitionElementToWomWorkflowDefinition.convertGraphElements(graphLikeConvertInputs)
+
+      innerGraph map { ig =>
+        val withOutputs = WomGraphMakerTools.addDefaultOutputs(ig, compoundCallIdentifiers = false)
+        println(s"nodes in this scatter: ${withOutputs.nodes.map(n => n.getClass.getSimpleName + " " + n.localName)}")
```
Comment (cjllanwarne) [inline wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/wdlom2wom/graph/ScatterElementToGraphNode.scala]: 🤦‍♂️

**Comment by geoffjentry** [comment]
@mcovarr I think the intention is to start moving people ASAP to WDL 1.0 once it's released next month. We'll need to support workflows which predate that, but it doesn't mean that we need to give those users new functionality and thus reasons to not upgrade

**Comment by mcovarr** [comment]
@geoffjentry yes I was more suggesting we don't allow nested scatters pre-version 3 🙂

**Comment by cjllanwarne** [comment]
@mcovarr @geoffjentry I agree.

FWIW if draft 2 ends up going through `wdlom` as well instead of having its own object model, it might pick up nested scattering for free. But (at least for now) I don't intend to retrofit this transform into the existing draft 2 to WOM conversion.

File: wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/wdlom2wom/graph/ScatterElementToGraphNode.scala [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+  def convertOuterScatter(a: ScatterNodeMakerInputs): ErrorOr[Set[GraphNode]] = {
+    val scatterExpression = a.node.scatterExpression
+    val scatterVariableName = a.node.scatterVariableName
+    val graphElements = a.node.graphElements
+
+    val scatterWomExpression: WdlomWomExpression = WdlomWomExpression(scatterExpression, a.linkableValues)
+    val scatterExpressionNodeValidation: ErrorOr[AnonymousExpressionNode] = AnonymousExpressionNode.fromInputMapping(WomIdentifier(scatterVariableName), scatterWomExpression, a.linkablePorts, PlainAnonymousExpressionNode.apply)
+
+    val scatterVariableTypeValidation: ErrorOr[WomType] = scatterExpression.evaluateType(a.linkableValues) flatMap {
+      case a: WomArrayType => a.memberType.validNel
```
[the remark this answers is not in the document]
Comment (cjllanwarne) [reply wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/wdlom2wom/graph/ScatterElementToGraphNode.scala]: oh, good point!

**Comment by cjllanwarne** [comment]
@Horneth yeah, I fixed it up in "Give scatters stable names so that they can be reconnected to as subw..."

**Comment by TMiguelT** [comment]
I note that this was merged more than 6 months ago, but the Cromwell docs still say
> Cromwell does not support nested scatters.

(https://github.com/broadinstitute/cromwell/blob/develop/docs/LanguageSupport.md#wdl-draft-2)

Is this still the case?

**Comment by cjllanwarne** [comment]
Thanks @TMiguelT - I've made https://github.com/broadinstitute/cromwell/issues/4224 to fix that page.

The short answer for you is that nested scatters do work but you'll need to be using WDL 1.0

**Pull request merged**